### ISSUE TITLE
inbound messages corrupt 

### ISSUE THREAD

**Issue by minhoca83** [knolleary/pubsubclient#834]
Hi everyone,
I have a project to turn On and Off lights remotely. 
I'm using a Arduino Mega and Sim5360 and want use MQTT (Mosquitto) as  comunication protocol. 
I have th following problem, when I send a message to a topic which my project is suscribed  and sustract the payload from the callback fuction great parto of the message is corrupt. 
I'm using the most recent Pubsubclient and TinyGSM librarys. 
I already check that I'm recived the MQTT payload right from the GSM module sfnfing in the UART comunication between the MCU and  the GSM module. 

For example this is the message  that I extrac from the GSM module snifing the UART comunication:
----------------------------------------------------------------------------------------------
AT+CIPRXGET=2,0,57
+CIPRXGET: 2,0,57,0
07VKA/sinapsis/MOSI{"ID":"868959031794765","CMD":"2"}

OK
----------------------------------------------------------------------------------------------

This is the answser and payload received from my callback fuction:
----------------------------------------------------------------------------------------------
Entra msj de MQTT
Message arrived [vikua/sinapsis/MOSI] 
 len = 34
(char)payload[]= {"ID":"86895903174"M2
Kÿÿÿÿÿ{FF}

deserializeJson() failed with code InvalidInput
------------------------------------------------------------------------------------------------

And part of my my callback fuction:

-----------------------------------------------------------------------------------------------
void mqttCallback(char* topic, byte* payload, unsigned int len) {
    #ifdef DEBUG_MQTT
    SerialMon.println(F("Entra msj de MQTT"));
    SerialMon.print("Message arrived [");
    SerialMon.print(topic);
    SerialMon.println("] ");
    SerialMon.print(" len = ");
    SerialMon.println(len);
    SerialMon.print("(char)payload[]= ");
   for (int i=0;i<len;i++) {
   SerialMon.print((char)payload[i]);
   }
    SerialMon.println();
    //SerialMon.print(F("Mensaje de data[i]: "));
    #endif
  //Compara si el topico es igual al que tiene
  if (String(topic) == MQTT.topicLed) {
    StaticJsonDocument<200> jsonBuffer;
    char data[len];
    for (int i=0;i<len;i++){
      data[i]=(char)payload[i];
      //#ifdef DEBUG_MQTT
      //SerialMon.print(data[i]);
      //#endif
    }
    #ifdef DEBUG_MQTT
    SerialMon.println();
    #endif
  // chequea si hay un error en la deserializacion del json
  auto error = deserializeJson(jsonBuffer, data);
  if (error){
    #ifdef DEBUG_MQTT
    SerialMon.print(F("deserializeJson() failed with code "));
    SerialMon.println(error.c_str());
    #endif
    return;
  }
  //Chequea si el json no esta vacio
  JsonObject root = jsonBuffer.as<JsonObject>();
  if (root.isNull()) {
    #ifdef DEBUG_MQTT
    SerialMon.println("parseObject() failed");
    SerialMon.println("Mensaje erroneo");
    #endif
  return;
  }
...
-------------------------------------------------------------------------

Most of the time I extrac corrupt messages but some times works prefectly and then bad again. 
I'm little lost where could be the buffer is ovewriting  Or coul be a compatibility problem betwen both librarys. 

Thanks in advance for any help

**Comment by TD-er** [comment]
Why don't you copy the topic and payload to the buffer first and then try to print it via serial?
Also what size do you use for the PubSubClient buffer?

**Comment by minhoca83** [comment]
Hi @TD-er, thank you very much for your replay,

Sorry but I would like assure that I not asumig wrong. So, are you sugest to create a new buffer in my callback fuction and fillet whith the topic and payload?  You don't believe that the problem could come from the library?

Anyway, I will try your suggestion. 

The size of the  PubSubClient buffer is the default:  #define MQTT_MAX_PACKET_SIZE 256

**Comment by TD-er** [comment]
Problem is that sending to serial may be blocking if you exceed the hardware buffer.
So by copying the data first you make sure you're comparing the actual data and not trying to find a cause which may be cause by something else.

So it is not a permanent solution, but more something to help diagnosing the issue.

**Comment by minhoca83** [comment]
I tried your suggestion but I got the same result...

About what you are telling me, why the hardware buffer could be involved in my callback function? if the reception of data through UART was already done by the library (TinyGSM).

Finally, I have decided to start over from the simple and I will try again with the example sketch of MqttClient from TinyGSM Library.  And I will find once for all if it's my code or the library.

Thank you very much for the suggestion and information. Any new information or other suggestions are all welcome!
I will update my results.

**Comment by TD-er** [comment]
You're sending data to the serial port to show you what you receive, right?
Not sure what size your serial buffer size is.
On an ESP8266 it is 128 bytes, but I can imagine it may differ per platform.
If you sent more data to the serial port than fits in this buffer, it will probably wait till it can fit all remaining data into the serial buffer.

But it may receive data (or need to receive data) inbetween, causing issues in receiving data.
So your debugging (sending to serial) may actually affect the process of receiving data.

**Comment by minhoca83** [comment]
Hi @ TD-er I have been keep digin and I think you are right about the size of the buffer because I discarded that the corruption is made on the PubSubClient Library. I went until TinyGSM library specifically to the fuction that reads the message recieve by the GSM module (ModemRead) and there I could see that while the fuction is extracting from the 40ths byte the code slow down and the data recive was any stranger ASCCI symbol.
By the way, I tested it without debougging or print anything ini the serial monitor and the problem persist ...
So, thank you very much for feedback and I will close this issue and open other in the TinyGSM repository.
